sympy/solvers/tests/test_solvers.py

```diff
@@ -1460,12 +1459,12 @@ def test_issue_5901():
     assert solve([f(x) - 3*f(x).diff(x)], f(x)) == \
         {f(x): 3*D}
     assert solve([f(x) - 3*f(x).diff(x), f(x)**2 - y + 4], f(x), y) == \
-        [{f(x): 3*D, y: 9*D**2 + 4}]
+        [(3*D, 9*D**2 + 4)]
```

Comment:
Why has this changed?

I don't really understand how this is decided:
```
In [1]: solve([x, y], x, y)
Out[1]: {x: 0, y: 0}

In [2]: solve([x**2, y], x, y)
Out[2]: {x: 0, y: 0}

In [3]: solve([x**2-1, y], x, y)
Out[3]: [(-1, 0), (1, 0)]

```

Reply:
If a system has a single solution it comes back as a dict, otherwise as a list of tuples *if solutions for all symbols have been found*.
```python
>>> solve([x**2-1, y])  # can't return tuple or else result is ambiguous
[{x: -1, y: 0}, {x: 1, y: 0}]
```
That's only partially true, however, as this example shows. In the case of `x**2, y` the system is split into two and the final result is a single solution so the splitter code returns the dict
```
            # Return a list with one dict as just the dict
            if len(sols) == 1:
                return sols[0]
            return sols
```
There is no communication that the single equation `[x**2]` was non-linear, only that it had a single solution.

I the case of the changed code, since the system was coupled and non-linear, the solution came back in tuple form indicating that the system had, potentially, more than one solution. Since a dictionary result was not requested, a list of dictionaries should not be returned. 

Current behavior should be tuple of solutions or single dictionary if symbols are provided, else single dictionary or list of dictionaries if symbols are not provided.

Reply:
I honestly can't wrap my head around what the correct return value should be.

How difficult/breaking would it be to change solve so that the type of the return value is determined purely by the types of the arguments?

Reply:
This was discussed at https://groups.google.com/g/sympy/c/v_YLkX4QuRY (that was the thread I was looking for earlier)

I think current behavior should be

- single equation, single variable: list of expressions
- single equation, 0 or more than 1 variable: list of dicts (*)
- list of equation(s), single solution -> dict
- list of equation(s), multiple solutions, ordered symbols with solutions for each: list of tuples
- list of equation(s), multiple solutions, unordered symbols or fewer solutions than symbols: list of dicts

(*) handling undetermined coefficients throws the wrench into this, however, because it will return a dict. But if that is removed, then the above scheme should hold.

And if one doesn't want that, select `dict=True` to always get a list of dictionaries.

Reply:
So isn't this case 3?

Reply:
Yes. I think we should minimize the dictionary return type, however. This is akin to returning a single expression instead of a singleton list if a single solution to one equation is found. So if we can limit the bare dictionary to a fully linear system that would be better. In that case, case 3 should read "list of *linear* equations" and give dict or [] (if there is no solution).

Reply:
Honestly none of this makes any sense to me at all. I can't imagine that it makes much sense to anyone else either but I assume that people will have written code that depends on the particular return types that they saw when calling solve so that any change now would break something for someone.

I'm just trying to work out if there's any way to make this more predictable or consistent without breaking too much. The most problematic cases are where the users passes exactly the same type of inputs but gets different types of outputs depending the value of the inputs. Distinguishing between linear and nonlinear is precisely the kind of thing that we should not do here.

It should be possible without looking at the equations themselves to say what the form of the output will be. In this case we have a list of two equations and two unknowns given explicitly. Can there be a unique rule that describes what the return type should be when there are two equations with two unknowns?

Reply:
Least breaking, I think, would be two types. A single equation and single variable: list of expressions; multiple equations or multiple symbols, list of dictionaries. But retain existing option to request output of dict=True to always give a list of dicts. Anything else can be handled at the user level to convert to other types.

Reply:
I think I've come to the conclusion that it's best not to change anything until there is a fully worked out new API. It could be called `solve2` or something but until we have that I think that the return types from `solve` should just be left alone.

Reply:
But what about `solve(eqs, syms)` where `eqs=[x - 1]` and `syms=[x]` or `solve(eqs, x)`? It's probably easiest to determine by inspection what the result is going to be by considering the variable. Hopefully by name or close context it is clear if one is dealing with a symbol or a container of symbols. The former could give the list of expressions as the result and the other, the list of dictionaries regardless of how the equation is passed. So the rule could be: 1) variable is a symbol -> list of expressions; 2) symbol(s) passed as a list -> list of dictionaries.

Reply:
> not to change anything

In the case of the changed doctest this should be considered wrong. It (and the correction) are inconsistent with the existing behavior in the list 1-5 above.

Reply:
Here is the current logic for variables passed as symbols:
```
if equation is a boolean expression or a list contains a boolean
	boolean value or expression
		>>> solve((x**2-4, x>1))
		Eq(x, 2)
		>>> solve((x**2-4, x>2))
		False

else expression or list of expression(s) is passed
    and there is no solution
		return []
	else
		equation is single expression
			0 symbols
				univariate - list of values
				else - list of dictionaries
			1 symbol
				list of values
			> 1 symbol
				list of tuples
		list of 1 or more expressions for equation with
			0 symbols
				linear in all variables - dict
				else - list of dictionaries
			1 symbol
				single solution - dict
				else - list of tuples
			> 1 symbol
				linear system and #eqs = #symbols - dict
				else - list of tuples
```

Reply:
If you wanted to flesh this out, this would make a good document for the "explanations" section of the documentation.

Reply:
Can you point me to a document or do you mean solve docstring, @asmeurer

Reply:
Got it: [doc/src/explanation/index.rst](https://github.com/sympy/sympy/pull/23746/files#diff-7bd55f350b73709e128b7a14c1dcc87f1a15685ca61eda91deb5616f4c498855)

Reply:
Yes that. Look at my improved descriptions of the sections at https://github.com/sympy/sympy/pull/23669. The "explanations" section is where we can go into depth to document things like implementation details and discuss design decisions.

Reply:
And also see https://diataxis.fr/

Reply:
> And also see

NICE! Glad to see that you are using that as a guide.

Reply:
When #23669 and this PR are merged I will begin deprecation of `solve` trying to do undetermined coefficients and add the explanation text similar to what I posted at https://groups.google.com/g/sympy/c/v_YLkX4QuRY .

Reply:
If there is no objection, this can go in on Monday.